cache.h

```diff
@@ -20,10 +20,10 @@
 #include <zlib.h>
 typedef struct git_zstream {
 	z_stream z;
-	unsigned long avail_in;
-	unsigned long avail_out;
-	unsigned long total_in;
-	unsigned long total_out;
+	size_t avail_in;
+	size_t avail_out;
+	size_t total_in;
+	size_t total_out;
```

Comment:
I think that these four changes are incorrect. The structure should have the same signature as the zlib library http://www.zlib.net/zlib_faq.html#faq32. 
That said, The consequential calculations that use these variables should be carefully checked to ensure that they are up cast to size_t or ptrdiff_t when pointer based calculations are being performed. 

On Windows the intermediate calculations for normal arithmetic expressions are still in long, not size_t. For the latter (pointer calcs) all the items in the expression will need to be upcast to memsized types, or at least that's my understanding of 'implementation defined' parts of this debacle.

Reply:
I think that this is good for `cache.h`, the details of the zlib adapter should be hidden in our `zlib.c`. For example, when `z_stream` cannot handle all of the `avail_in` provided by the caller in Git's source code, then `zlib.c` should take the pain away from the caller by doing the obvious thing: feed things to zlib in a loop.

Reply:
at the Dev Conf @t-b and I had a look at one of the bits of code that t-b had identified that was failing. It was the bit of code that Junio had tentatively put in to handle small memory footprints and hopefully the Windows issues by doing what you say. However it wasn't working for Windows but does for Linux. 

If I understood t-b's setup, he already had the patches in this PR, but was still stuck on the https://github.com/gitgitgadget/git/blob/1ec976b25f74f986eedce4c006d994a3d3ed9431/zlib.c#L52-L57 code.

It is expected (guessed) that we have a coercion problem here, in that zlib itself is only 32bit aware for byte counting (on Windows - FAQ32), but is happily 64 bit aware for the pointers.
I've been trying to get my setup back to scratch (see PRs various) so that I can also have a test of the ideas, and ensuring I have fingered the correct parts (i.e. I could have pointed at the wrong part of the struct..).

Reply:
Okay, then. Let me know if you want me to look at anything specifically.

Reply:
The changes to cache.h are IMHO correct. We need to be able to inflate/deflate files which are larger than 2^32. The quoted lines from @PhilipOakley are something which needs to be fixed as part of the  zlib unsigned long removal switch in #116. More upstream discussion is at https://public-inbox.org/git/994568940.109648.1548957557643@ox.hosteurope.de/.

I'm posting that mainly here now as I will not be able tp pursue this task further in the next few months. I'm eager to get that supported since 10 years so there is no point in rushing now ;)